grails-app/services/org/pih/warehouse/importer/InventoryImportDataService.groovy

```diff
@@ -128,74 +138,172 @@ class InventoryImportDataService implements ImportDataService {
 
     @Override
     void importData(ImportDataCommand command) {
-        def dateFormatter = new SimpleDateFormat("yy-mm")
+        InventoryImportData inventoryImportData = parseData(command)
+
+        Map<String, AvailableItem> availableItems = productAvailabilityService.getAvailableItemsAtDateAsMap(
+                command.location,
+                new ArrayList<Product>(inventoryImportData.products),
+                command.date)
+
+        // Date objects are mutable, so we use Instant to clone the date in the command and avoid directly modifying it.
+        // -1 (instead of +1 to the adjustment transaction date) because transaction date can't be in the future.
+        Date baselineTransactionDate = DateUtil.asDate(DateUtil.asInstant(command.date).minusSeconds(1))
 
-        def transaction = new Transaction()
-        transaction.transactionDate = command.date
-        transaction.transactionType = TransactionType.get(Constants.PRODUCT_INVENTORY_TRANSACTION_TYPE_ID)
+        String comment = "Imported from ${command.filename} on ${new Date()}"
+
+        inventoryImportProductInventoryTransactionService.createInventoryBaselineTransactionForGivenStock(
+                command.location, command, availableItems.values(), baselineTransactionDate, comment)
+
+        createAdjustmentTransaction(command.location, inventoryImportData, availableItems, command.date, comment)
+    }
+
+    /**
+     * Create the adjustment transaction based on the difference between the QoH in the system at the date specified
+     * by the import and the quantity specified in the import.
+     */
+    private Transaction createAdjustmentTransaction(Location facility,
+                                                    InventoryImportData inventoryImportData,
+                                                    Map<String, AvailableItem> availableItems,
+                                                    Date transactionDate=null,
+                                                    String comment=null) {
+
+        // Don't bother populating the transaction's fields until we know we'll need one.
+        Transaction transaction = new Transaction()
+
+        for (entry in inventoryImportData.rows) {
+            String key = entry.key
+            InventoryImportDataRow row = entry.value
+
+            // If we have no product availability record for the item, we assume that means we have no quantity.
+            int quantityOnHand = availableItems.get(key)?.quantityOnHand ?: 0
```

Comment:
@alannadolny this is what I was mentioning in your PR. Because we have a map that is keyed on product + bin + lot we can just do a lookup on that key to fetch the QoH instead of looping all the AvailableItems.